### ISSUE TITLE
Ensure test circuits can be compiled in parallel

### ISSUE THREAD

**Issue by ivokub** [Consensys/gnark#195]
See discussion [here](https://github.com/ConsenSys/gnark/pull/191#discussion_r765802476)

**Comment by gbotrel** [comment]
That is; **compiling** the same circuit in parallel is an issue, but I think mostly for internal testing. In a normal workflow, I don't see why a circuit developer would want to compile the exact same instance of a circuit more than 1 time, and in parallel (?).

**Comment by ivokub** [comment]
> That is; **compiling** the same circuit in parallel is an issue, but I think mostly for internal testing. In a normal workflow, I don't see why a circuit developer would want to compile the exact same instance of a circuit more than 1 time, and in parallel (?).

Makes sense, there shouldn't be a plausible reason to compile the same circuit in parallel outside of testing infrastructure. But I guess within testing it is reasonable to have a lock on the circuit (in the TestCircuit structure) as the circuits are used in different tests and possibly compiled in parallel.

I guess this makes this issue less relevant, but still nice-to-have in the future.
